src/median.sol

```diff
@@ -0,0 +1,129 @@
+
+import {IMedian} from "./interfaces/IMedian.sol";
+
+contract Median is IMedian, Initializable, OwnableUpgradeable, UUPSUpgradeable {
+    using MessageHashUtils for bytes;
+    using ECDSA for bytes32;
+
+    bytes32 public constant currencyPair = 0x555344432f784e474e0000000000000000000000000000000000000000000000; // hex("USDC/xNGN);
+
+    uint32 public minimumQuorum;
+    uint64 internal lastTimestamp;
+    uint128 internal lastPrice;
+
+    mapping(address => bool) public authorizedRelayers;
+    uint32 public authorizedRelayersCount;
+
+    PriceData[] public priceHistory;
+
+    function initialize(uint32 _minimumQuorum) public initializer {
+        __Ownable_init_unchained(msg.sender);
+
+        minimumQuorum = _minimumQuorum;
+    }
+
+    function _authorizeUpgrade(address newImplementation) internal override onlyOwner {}
+
+    modifier onlyAuthorizedRelayer() {
+        if (!authorizedRelayers[msg.sender]) revert OnlyAuthorizedRelayers();
+
+        _;
+    }
+
+    modifier hasMinimumQuorum(uint256 pricesLength) {
+        if (pricesLength < minimumQuorum) revert NotEnoughPrices();
+        _;
+    }
+
+    function authorizeRelayer(address _relayerAddress) external onlyOwner {
+        if (!authorizedRelayers[_relayerAddress]) {
+            authorizedRelayers[_relayerAddress] = true;
+            authorizedRelayersCount++;
+        }
+
+        emit AuthorizedRelayer(_relayerAddress);
+    }
+
+    function deauthorizeRelayer(address _relayerAddress) external onlyOwner {
+        if (authorizedRelayers[_relayerAddress]) {
+            authorizedRelayers[_relayerAddress] = false;
+            authorizedRelayersCount--;
+        }
+
+        emit DeauthorizedRelayer(_relayerAddress);
+    }
+
+    function updateMinimumQuorum(uint32 _minimumQuorum) external onlyOwner {
+        if (_minimumQuorum == 0) revert InvalidQuorum();
+        minimumQuorum = _minimumQuorum;
+
+        emit MinimumQuorumUpdated(_minimumQuorum);
+    }
+
+    function read() external view returns (uint256, uint256) {
+        return (lastTimestamp, lastPrice);
+    }
+
+    function update(uint256[] calldata _prices, uint64[] calldata _timestamps, bytes[] calldata _signatures)
+        external
+        onlyAuthorizedRelayer
+        hasMinimumQuorum(_prices.length)
+    {
+        if (_prices.length != _timestamps.length || _prices.length != _signatures.length) revert InvalidArrayLength();
+
+        // cache timestamp on the stack to save gas
+        uint256 _lastTimestamp = lastTimestamp;
+        uint256 _trackedPrice;
+
+        for (uint256 i; i < _prices.length; ++i) {
+            if (_timestamps[i] <= _lastTimestamp) revert InvalidTimestamp();
+            if (_prices[i] < _trackedPrice) revert PricesNotOrdered();
+
+            address signer = recover(_prices[i], _timestamps[i], currencyPair, _signatures[i]);
+
+            // ecdsa lib already reverts with error `ECDSAInvalidSignature()` if signer == address(0)
```

Comment:
Make sense, read somewhere that In some cases ecrecover can return a random address instead of 0 for an invalid signature.

Reply:
Yeah most times it always recovers to an address, it only returns address(0) if r is not a valid x axis point on the curve, if v is not 27 or 28 and if the public key is the identity element of the curve. If not it'll recover to a valid public key with a valid address. 

That's why there's another check to ensure signer is msg.sender (even though the open zeppelin Lib reverts if it returns address 0)